### PR TITLE
adding `accelerate-deepspeed` blog

### PR DESCRIPTION
**Pull request by pacman100**

### CODE REVIEW

File: accelerate-deepspeed.md [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+    <a href="/sgugger">
+        <img class="avatar avatar-user" src="https://aeiljuispo.cloudimg.io/v7/https://s3.amazonaws.com/moonup/production/uploads/1593126474392-5ef50182b71947201082a4e5.jpeg?w=200&h=200&f=face" title="Gravatar">
+        <div class="bfc">
+            <code>sgugger</code>
+            <span class="fullname">Sylvain Gugger</span>
+        </div>
+    </a>
+</div>
+
+In this post we will look at how we can leverage **[Accelerate](https://github.com/huggingface/accelerate)** Library for training large models which enables users to leverage ZeRO features of **[DeeSpeed](https://www.deepspeed.ai)**.
```
Comment (sgugger) [inline accelerate-deepspeed.md:32]:
```suggestion
In this post we will look at how we can leverage the **[Accelerate](https://github.com/huggingface/accelerate)** library for training large models which enables users to leverage the ZeRO features of **[DeeSpeed](https://www.deepspeed.ai)**.
```

File: accelerate-deepspeed.md [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+
+# Motivation 🤗
+**Tired of OOM errors while trying to train large models? We have got you covered. Large models are very performant [1] but difficult to train with the available hardware. To get the most of the available hardware for training large models one can leverage Data Parallelism using ZeRO - Zero Redundancy Optimizer [2]**.
```
Comment (sgugger) [inline accelerate-deepspeed.md:35]:
```suggestion
**Tired of OOM errors while trying to train large models? We've got you covered. Large models are very performant [1] but difficult to train with the available hardware. To get the most of the available hardware for training large models one can leverage Data Parallelism using ZeRO - Zero Redundancy Optimizer [2]**. 
```

File: accelerate-deepspeed.md [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+---
+| Method | Batch Size Max | Train time per epoch (seconds) | Eval time  per epoch (seconds) | F1 score | Accuracy |
+| --- | --- | --- | --- | --- | --- |
+| DDP (Distributed Data Parallel) | 8 | 103.57 | 2.04 | 0.931 | 0.904 |
+| DeepSpeed ZeRO Stage 2 | **40** | **28.98** | **1.79** | **0.936** | **0.912** |
+
+Table 1: Benchmarking DeepSpeed ZeRO Stage-2 on DeBERTa-XL (900M) model
+
+---
+We observe ~**3.5X** speed up in training without any drop in perforamnce metrics, all this without changing any code. Yay! 🤗.
```
Comment (sgugger) [inline accelerate-deepspeed.md:114]:
```suggestion
With this bigger batch size, we observe ~**3.5X** speed up in total training time without any drop in perforamnce metrics, all this without changing any code. Yay! 🤗. 
```
(Just so people don't believe it's the training step that is faster.)

File: accelerate-deepspeed.md [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+| --- | --- | --- | --- | --- | --- |
+| DDP (Distributed Data Parallel) | 8 | 103.57 | 2.04 | 0.931 | 0.904 |
+| DeepSpeed ZeRO Stage 2 | **40** | **28.98** | **1.79** | **0.936** | **0.912** |
+
+Table 1: Benchmarking DeepSpeed ZeRO Stage-2 on DeBERTa-XL (900M) model
+
+---
+We observe ~**3.5X** speed up in training without any drop in perforamnce metrics, all this without changing any code. Yay! 🤗. 
+
+To be able to tweak more options, you will need to use a DeepSpeed config file and minimal code changes. Next, We will see how to do this.
```
Comment (sgugger) [inline accelerate-deepspeed.md:116]:
```suggestion
To be able to tweak more options, you will need to use a DeepSpeed config file and minimal code changes. Let's see how to do this.
```

File: accelerate-deepspeed.md [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+
+# Accelerate 🚀:  Leverage DeepSpeed Config file to tweak more options
```
Comment (sgugger) [inline accelerate-deepspeed.md:118]:
```suggestion
# Accelerate 🚀:  Leverage a DeepSpeed Config file to tweak more options
```

File: accelerate-deepspeed.md [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+
+First, We will look at the task of finetuning Sequence-to-Sequence model for training our own Chatbot. Specifically, we will finetune `facebook/blenderbot-400M-distill` on `smangrul/MuDoConv` (Multi-Domain Conversation) dataset. The dataset contains conversations from 10 different data sources convering personas, grouding in specific emotional contexts, goal-oriented (e.g., restaurant reservation) and general wikipedia topics (e.g, Cricket).
```
Comment (sgugger) [inline accelerate-deepspeed.md:120]:
```suggestion
First, We will look at the task of finetuning a sequence-to-sequence model for training our own Chatbot. Specifically, we will finetune `facebook/blenderbot-400M-distill` on the `smangrul/MuDoConv` (Multi-Domain Conversation) dataset. The dataset contains conversations from 10 different data sources covering personas, grounding in specific emotional contexts, goal-oriented (e.g., restaurant reservation) and general wikipedia topics (e.g, Cricket).
```
+ link to the dataset on the Hub!

File: accelerate-deepspeed.md [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+
+
+The code is available here [run_seq2seq_no_trainer.py](https://github.com/pacman100/accelerate-deepspeed-test/blob/main/src/modeling/run_seq2seq_no_trainer.py). Current pratice to effectively measure the `Engagingness` and `Humanness` of Chatbots is via Human evlauations which are expensive [6]. As such for this example, the metric being tracked is BLEU score which isn't ideal but conventional automatic metric for such tasks. One can adapt the code to train larger T5 models if you have access to GPUs that support `bfloat16` precision else you will run into `NaN` loss values. We will run a quick benchmark on `10000` train samples and `1000` eval samples as we are interested in DeepSpeed vs DDP.
```
Comment (sgugger) [inline accelerate-deepspeed.md:123]:
```suggestion
The code is available here [run_seq2seq_no_trainer.py](https://github.com/pacman100/accelerate-deepspeed-test/blob/main/src/modeling/run_seq2seq_no_trainer.py). Current pratice to effectively measure the `Engagingness` and `Humanness` of Chatbots is via Human evlauations which are expensive [6]. As such for this example, the metric being tracked is BLEU score (which isn't ideal but is the conventional metric for such tasks). One can adapt the code to train larger T5 models if you have access to GPUs that support `bfloat16` precision else you will run into `NaN` loss values. We will run a quick benchmark on `10000` train samples and `1000` eval samples as we are interested in DeepSpeed vs DDP.
```

File: accelerate-deepspeed.md [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+    "gradient_accumulation_steps": 1,
+    "gradient_clipping": "auto",
+    "steps_per_print": 2000,
+    "train_batch_size": "auto",
+    "train_micro_batch_size_per_gpu": "auto",
+    "wall_clock_breakdown": false
+}
+```
+
+To enable DeepSpeed ZeRO Stage-2 with above config, please run `accelerate config` and provide the config file path when asked. For more details, refer the 🤗 `accelerate` official documentation [DeepSpeed Config File](https://huggingface.co/docs/accelerate/deepspeed#deepspeed-config-file).
```
Comment (sgugger) [inline accelerate-deepspeed.md:172]:
```suggestion
To enable DeepSpeed ZeRO Stage-2 with above config, please run `accelerate config` and provide the config file path when asked. For more details, refer the 🤗 `accelerate` official documentation for [DeepSpeed Config File](https://huggingface.co/docs/accelerate/deepspeed#deepspeed-config-file).
```

File: accelerate-deepspeed.md [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+- lr_scheduler = get_scheduler(
+-            name=args.lr_scheduler_type,
+-            optimizer=optimizer,
+-            num_warmup_steps=args.num_warmup_steps,
+-            num_training_steps=args.max_train_steps,
+-        )
+
++ lr_scheduler = DummyScheduler(
++            optimizer, total_num_steps=args.max_train_steps, warmup_num_steps=args.num_warmup_steps
++        )
```
Comment (sgugger) [inline accelerate-deepspeed.md:236]: Make sure the indentation is just 4 spaces here.

File: accelerate-deepspeed.md [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+
+![Chatbot](./assets/82_accelerate_deepspeed/chatbot.png)
+
+---
+## CPU/Disk Offloading to enable training humongous models that won’t fit the GPU memory
+
+On a single 24GB NVIDIA Titan RTX GPU, one cannot train GPT-XL Model (1.5B parameters) even with a batch size of 1. We will look at how we can use DeepSpeed ZeRO Stage-3 with CPU offloading of optimizer states, gradients and parameters to train GPT-XL Model. 
+
+We will leverage the DeepSpeed Zero Stage-3 CPU offload config [zero3_offload_config_accelerate.json](https://github.com/pacman100/accelerate-deepspeed-test/blob/main/src/modeling/configs/zero3_offload_config_accelerate.json) (given below) For training. The rest of the process of using the config with 🤗 `accelrate` is similar to above experiment.
```
Comment (sgugger) [inline accelerate-deepspeed.md:256]:
```suggestion
We will leverage the DeepSpeed Zero Stage-3 CPU offload config [zero3_offload_config_accelerate.json](https://github.com/pacman100/accelerate-deepspeed-test/blob/main/src/modeling/configs/zero3_offload_config_accelerate.json) (given below) for training. The rest of the process of using the config with 🤗 `accelerate` is similar to the above experiment.
```

**Review by sgugger — APPROVED** [review]
Nice post, thanks a lot for writing it! Left a couple of nits.
